File: benchmarks/pandas/bench_interpolate_bfill_limit.py
"""
Benchmark: Series.interpolate with bfill method and limit option — backward fill with gap limit on 50k Series.
Outputs JSON: {"function": "interpolate_bfill_limit", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 50_000
WARMUP = 5
ITERATIONS = 30

data = np.where(np.arange(SIZE) % 7 < 2, np.nan, np.sin(np.arange(SIZE) * 0.01) * 100)
s = pd.Series(data)

for _ in range(WARMUP):
    s.interpolate(method="bfill")
    s.ffill(limit=2)
    s.bfill(limit=1)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.interpolate(method="bfill")
    s.ffill(limit=2)
    s.bfill(limit=1)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "interpolate_bfill_limit",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))

File: benchmarks/tsb/bench_interpolate_bfill_limit.ts
/**
 * Benchmark: interpolateSeries with bfill method and limit option — backward fill with gap limit on 50k Series.
 * Outputs JSON: {"function": "interpolate_bfill_limit", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, interpolateSeries } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

// ~15% NaN values with consecutive gaps of up to 5
const data = Array.from({ length: SIZE }, (_, i) => {
  const gap = i % 7;
  if (gap === 0 || gap === 1) return null;
  return Math.sin(i * 0.01) * 100;
});
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  interpolateSeries(s, { method: "bfill" });
  interpolateSeries(s, { method: "ffill", limit: 2 });
  interpolateSeries(s, { method: "bfill", limit: 1 });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  interpolateSeries(s, { method: "bfill" });
  interpolateSeries(s, { method: "ffill", limit: 2 });
  interpolateSeries(s, { method: "bfill", limit: 1 });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "interpolate_bfill_limit",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
